Enforce PSR-12 and organise imports on Winter CMS core files

bennothommo opened the pull request:
This commit changes the PHP code sniffer to use PSR-12 rules, as that ruleset will pick up a lot of the code style issues being flagged in PRs recently.

In order to set the baseline, all current core files have been reformatted to fit the ruleset.

I have also implemented rules around imports in files - mainly, taking out the aliases and using qualified facades and class names, and stripping out unused imports. This will improve IDE support for PHP intelligence and code completion.

Arrays are also now enforced to have a trailing comma to improve diffs.

Sorry in advance for the code review... :grimacing:

- [x] Docs updates (https://github.com/wintercms/docs/pull/205)
- [x] Add `php artisan winter:phpcs` command
- [x] Ensure stubs follow correct style

bennothommo committed:
Enforce PSR-12 on Winter CMS core files

This commit changes the PHP code sniffer to use PSR-12 rules, as that ruleset will pick up a lot of the code style issues being flagged in PRs recently.

In order to set the baseline, all current core files have been reformatted to fit the ruleset.

bennothommo committed:
Additional code style fixes to improve future diffs and IDE support

- Convert all aliases to qualified class names (mainly facades)
- Reorder imports alphabetically
- Ensure last item in array has a trailing comma
- Strip out unused imports

bennothommo committed:
Use Composer-derived PHPCS, udpate GitHub Actions

bennothommo committed:
Update GitHub Actions

bennothommo committed:
Reset modules before checking code quality

LukeTowers commented on .github/workflows/manifest.yml:
```diff
@@ -74,13 +43,8 @@ jobs:
-      - name: Cancel previous incomplete runs
-        uses: styfle/cancel-workflow-action@0.8.0
-        with:
-          access_token: ${{ github.token }}
```
Are we not using caching / cancelling previous runs anymore?

bennothommo replied:
We are in some actions, using the new `concurrency` configuration. However, the manifest should probably complete even if it has been superseded, otherwise it could corrupt the manifest file.

LukeTowers commented on modules/backend/formwidgets/relation/partials/_relation.php:
```diff
@@ -1,3 +1,3 @@
 <div class="relation-widget" id="<?= $this->getId() ?>">
-    <?= $this->makePartial('~/modules/backend/widgets/form/partials/_field_'.$field->type.'.php', ['field' => $field]) ?>
+    <?= $this->makePartial('~/modules/backend/widgets/form/partials/_field_' . $field->type . '.php', ['field' => $field]) ?>
 </div>
```
@bennothommo do we have any rules for trailing newline on all files?

bennothommo replied:
Yeah I believe that's included in PSR12, but might not apply for the templates / partials.

LukeTowers replied:
Can we add a rule to make it apply?

bennothommo replied:
It appears not - the code sniffer can only look at PHP blocks, it cannot read HTML as far as I can tell

LukeTowers commented on modules/backend/layouts/_custom_styles.php:
```diff
@@ -1,6 +1,8 @@
+
     use Backend\Models\BrandSetting;
     use Backend\Models\EditorSetting;
+
```
Do we have to have this whitespace? It doesn't add much IMO

bennothommo replied:
No we don't, probably just a tweak to the config for one of the rules. I'll have a look.

bennothommo replied:
I've added an exception for that rule for partials, since changing the config changes it across the board and would allow the following in a PHP-only file (which looks wonky IMO):

```php
<?php
namespace Winter\Storm\Classes;
use Illuminate\Support\Facades\DB;
class Something extends SomethingElse
{
...
```

LukeTowers commented on modules/backend/widgets/MediaManager.php:
```diff
@@ -926,13 +927,15 @@ protected function getCurrentFolder(): string
             MediaLibraryItem::FILE_TYPE_DOCUMENT,
-            MediaLibraryItem::FILE_TYPE_VIDEO
-        ])) {
+            MediaLibraryItem::FILE_TYPE_VIDEO,
```
Should this be indented one more level?

bennothommo replied:
I believe that's another rule, but not one I'm using right now. Will take a look.

bennothommo replied:
Added rule.

LukeTowers commented on modules/backend/widgets/lists/partials/_list_head_row.php:
```diff
@@ -20,7 +20,7 @@
                     style="width: <?= $column->width ?>"
                 <?php endif ?>
-                class="<?= $this->sortColumn==$column->columnName?'sort-'.$this->sortDirection.' active':'sort-desc' ?> list-cell-name-<?= $column->getName() ?> list-cell-type-<?= $column->type ?> <?= $column->getAlignClass() ?> <?= $column->headCssClass ?>"
+                class="<?= $this->sortColumn == $column->columnName ? 'sort-' . $this->sortDirection . ' active' : 'sort-desc' ?> list-cell-name-<?= $column->getName() ?> list-cell-type-<?= $column->type ?> <?= $column->getAlignClass() ?> <?= $column->headCssClass ?>"
```
Should this be `===`?

bennothommo replied:
The original functionality was just the two `==`. It wouldn't change that, it would just ensure there's correct spacing around. I'd say leave it as is unless we're sure changing it to strict won't break anything.

LukeTowers commented on modules/cms/twig/SnowboardNode.php:
```diff
@@ -1,11 +1,13 @@
-use Config;
+namespace Cms\Twig;
+
+use Illuminate\Support\Facades\URL;
```
Any issues with the "URL" vs "Url"?

bennothommo replied:
Not that I have seen - I don't think PHP is case-sensitive in that regard.

LukeTowers commented on modules/system/models/MailLayout.php:
```diff
@@ -1,11 +1,13 @@
-use Model;
+namespace System\Models;
+
+use File as FileHelper;
```
Which File is this?

LukeTowers commented on modules/system/partials/_settings_menu.php:
```diff
@@ -1,4 +1,5 @@
 <?php
+
```
This one doesn't make any sense

bennothommo committed:
Exclude some rules for partials and views

bennothommo committed:
Add rule for array indentation

bennothommo committed:
Fix reference to File facade

LukeTowers commented:
@bennothommo can you submit a PR to our developer guidelines docs to reflect these changes? Also, is it worth adding a `winter:phpcs` command to run the coding styles against plugins and then providing that as a command that plugin devs can run against their codebase to make sure that it conforms with the community standards?

Another thing to check is our stub files, I'm not sure if they've all been updated to our latest preferences for generating PHP files.

bennothommo committed:
Merge branch 'develop' into fix/psr-12

bennothommo committed:
Split off Winter CMS code style ruleset to its own XML

This will allow us to use the ruleset in plugins as well

bennothommo committed:
Remove partial from exclusion list

With new ruleset, it now passes

bennothommo committed:
Add "winter:sniff" command

This command will allow developers to check code style against the Winter CMS code style rules in both the core and installed plugins.

LukeTowers commented on modules/system/console/WinterSniff.php:
```diff
@@ -0,0 +1,279 @@
+        parent::__construct();
+
+        // Register aliases for backwards compatibility with October
+        $this->setAliases(['winter:phpcs']);
```
We have a property for that already: https://github.com/wintercms/winter/blob/develop/modules/system/console/asset/mix/MixCompile.php#L34-L39

bennothommo committed:
Merge branch 'develop' into fix/psr-12

bennothommo committed:
Fix remaining code smells

bennothommo committed:
Change alias command

LukeTowers commented on phpcs.base.xml:
```diff
@@ -0,0 +1,184 @@
+            <property name="searchAnnotations" value="true" />
+        </properties>
+    </rule>
+</ruleset>
```
This file should probably live in the system module and be referenced from the stubs as such to avoid having the core ruleset changed on a per project basis and also to support existing projects making use of this feature when they don't have this file present since it's not distributed through a composer module update.

LukeTowers commented on composer.json:
```diff
@@ -42,6 +42,7 @@
         "mockery/mockery": "^1.4.4",
         "fakerphp/faker": "^1.9.2",
         "squizlabs/php_codesniffer": "^3.2",
+        "slevomat/coding-standard": "^8.15.0",
```
Is this required for the phpcs command to work? If so we should document it in the commit message / release notes.

LukeTowers replied:
And or maybe include it in the system module's composer.json?

bennothommo replied:
Yes, it is required for the extra rules we use in the coding standard. We can't include it in the system module however because you cannot include development requirements from a Composer dependency AFAIK. It's why we have a lot of duplicate development requirements in the main repo and Storm, for example.

LukeTowers committed:
Add support for `--fix` option to `sniff` command

LukeTowers committed:
Alias `winter:test` to `test`

LukeTowers commented:
Initial thoughts from testing on a client project:

## Problems:
- [x] Running on existing project with old phpcs.xml probably won't work because of the reference to winter.demo
- [x] Running on no file doesn't work because of the dependency on phpcs.base.xml being in the root (solve by moving to system module)
- [x] Old require-dev is insufficient to run as we require SlevomatCodingStandard rules
- [x] The scaffolded root phpcs.xml references demo plugin, doesn't exist on most existing projects, also wouldn't cover their plugins, so what is the purpose of a project root level file?
- [x] bootstrap/app.php, autoload.php fail on existing projects
- [x] Ton of warnings on core files from default configuration
- [x] No ability to add --fix flag to switch to phpcbf

## Coding style suggestions:
- [ ] Disallow or warn on loose type comparison (== vs ===)
- [x] Use of overridden Illuminate facades when a winter one exists
    - Config, DB, Event, File, Http (Do not block this one, Laravel one should be what we use going forwards), Mail?, Url?, Validator?
- [x] Block "use Model;" class alias instead of Winter\Storm\Database\Model (block or warn against all global use imports if possible, or make it easier for IDEs to play nice with them)
- [ ] Not using anonymous classes in migrations
- [ ] Empty methods in Plugin.php (or other clases, perhaps just issue a warning?)
- [ ] Use of routes in Plugin.php instead of in the routes.php file
- [ ] Semicolon on line by itself (usually unintentional, can result from a fix that applies ;;)
- [ ] Warning for .htm files in the backend view directories (example: controllers/*/.htm)
- [ ] Use of dot syntax for `implement` values (when not using soft implementation `@`). I.e. `Backend.Behaviors.FormController` should always be `\Backend\Behaviors\FormController::class`.
- [ ] Alphabetical order for all second+ level keys in translation files? (i.e. `*.*`, but root level keys can remain unalphabetized)
- [ ] Warning for `$formConfig = 'config_form.yaml'` etc (overridden property matches default value).

## Enhancement Ideas
- [x] Switch WinterTest to extend Winter's console command and alias `test` to it
- [x] Alias `sniff` to `winter:sniff`
- [x] Add support for `--fix` option in sniff
- [ ] Evaluate Laravel coding standards for alignment with Winter: https://docs.styleci.io/presets#laravel
- [ ] Add support for `--baseline` option in sniff to generate a baseline to ignore errors from
- [ ] Add support for module options to match `winter:test`
- [ ] Make winter VSCode extension auto load or recommend the PHPCS extension
- [ ] Add github action support for auto-fixing files on PRs
- [ ] Provide coding standard as an external repository so we can import it on our packages & plugins
- [x] Add composer scripts `test`, `sniff`, etc

## To test
- [x] Test running with the same package identifiers as test
- [x] Test running on a new plugin
- [x] Test phpcsfixer
- [x] Test output of all scaffold commands for adherance to standards
- [ ] Run sniff & fix on all first party Winter plugins and review changes / errors / warnings from those

LukeTowers committed:
Improve / relocate PHPCS stubs

LukeTowers committed:
Update composer scripts

LukeTowers committed:
Add class constant visibility

LukeTowers committed:
Fix remaining warnings

LukeTowers committed:
Add progress flag to phpcs

LukeTowers committed:
Add initial version of NoGlobalAliasesSniffTest

LukeTowers committed:
Add support for --parallel option to winter:sniff

LukeTowers committed:
Fix remaining global imports

LukeTowers committed:
Ignore the project's phpcs.xml file when exporting it as a new project

The core phpcs.xml is only relevant to core development, not projects using Winter. Running winter:sniff without a config file present will auto generate one from the stubs in the system module.

LukeTowers committed:
Fix overzealous find/replace

LukeTowers committed:
Remove debug code

LukeTowers committed:
Use Winter's Url facade instead of Laravel's

LukeTowers committed:
Typo fix

LukeTowers committed:
Allow use statements to reference classes in the same namespace

LukeTowers committed:
Fix default scaffolding stub files to conform to coding standard

bennothommo committed:
Merge branch 'develop' into fix/psr-12

LukeTowers committed:
Merge branch 'develop' into fix/psr-12

LukeTowers committed:
Merge branch 'develop' into fix/psr-12

LukeTowers committed:
Merge branch 'develop' into fix/psr-12

LukeTowers committed:
Update Parameter.php

LukeTowers committed:
Update PluginManager.php

LukeTowers committed:
Update WinterTest.php

LukeTowers committed:
Update EventLog.php

LukeTowers committed:
Update SecurityPolicyTest.php

LukeTowers committed:
Merge branch 'develop' into fix/psr-12

LukeTowers committed:
Merge branch 'develop' into fix/psr-12

# Conflicts:
#	modules/backend/behaviors/ImportExportController.php
#	modules/backend/classes/BackendController.php
#	modules/backend/classes/Controller.php
#	modules/backend/controllers/UserRoles.php
#	modules/backend/controllers/users/create.php
#	modules/backend/widgets/Filter.php
#	modules/backend/widgets/lists/partials/_list_head_row.php
#	modules/cms/classes/Controller.php
#	modules/system/console/WinterEnv.php
#	modules/system/console/WinterUtil.php
#	modules/system/controllers/maillayouts/create.php
#	modules/system/controllers/mailpartials/create.php
#	modules/system/controllers/mailtemplates/create.php
#	modules/system/models/MailSetting.php

LukeTowers committed:
Style fixes
